File: tests/help_test.ts
/**
 * Structural invariants for the help catalog (dev log 113, Phase 0).
 * Content-level help tests (router, rendering, facts) land in Phase 4.
 */
import { assertEquals } from "@std/assert";
import {
  HELP_PAGES,
  aliasOf,
  findByAlias,
  findByReach,
  findPage,
  groupOf,
  groupedPages,
  reachOf,
} from "../src/help.ts";
import { renderPage, renderPageRich, renderPageText } from "../src/help_render.ts";

/** Remove ANSI escape sequences. */
// deno-lint-ignore no-control-regex
const stripAnsi = (s: string): string => s.replace(/\x1b\[[0-9;]*m/g, "");

/** The 18-page page map from the help draft, exactly. */
const EXPECTED_IDS = [
  "home",
  "model/cst",
  "model/guarantees",
  "concepts/state-scope",
  "concepts/private-notes",
  "concepts/insets",
  "concepts/selectors",
  "concepts/mutations",
  "concepts/tracked-changes",
  "commands/init",
  "commands/schema",
  "commands/dump",
  "commands/read",
  "commands/bib",
  "commands/set",
  "commands/delete",
  "commands/insert",
  "commands/undo",
];

Deno.test("help catalog - page map matches the draft (18 pages)", () => {
  const ids = HELP_PAGES.map((p) => p.id).sort();
  assertEquals(ids, [...EXPECTED_IDS].sort());
});

Deno.test("help catalog - page IDs are unique", () => {
  const ids = HELP_PAGES.map((p) => p.id);
  assertEquals(new Set(ids).size, ids.length);
});

Deno.test("help catalog - reaches are unique (one canonical lq help <page> form)", () => {
  const reaches = HELP_PAGES.map((p) => reachOf(p.id));
  assertEquals(new Set(reaches).size, reaches.length);
});

Deno.test("help catalog - command aliases are unique and only on command pages", () => {
  const aliases = HELP_PAGES.map((p) => aliasOf(p.id)).filter((a): a is string => a !== undefined);
  assertEquals(new Set(aliases).size, aliases.length);
  // 9 command pages (init schema dump read bib set delete insert undo)
  assertEquals(aliases.length, 9);
});

Deno.test("help catalog - every page has a non-empty title", () => {
  for (const p of HELP_PAGES) assertEquals(p.title.length > 0, true);
});

Deno.test("help catalog - group/reach/alias are derived from id", () => {
  assertEquals(reachOf("commands/read"), "read");
  assertEquals(reachOf("model/cst"), "cst");
  assertEquals(reachOf("home"), "home");
  assertEquals(groupOf("commands/read"), "commands");
  assertEquals(groupOf("concepts/insets"), "concepts");
  assertEquals(groupOf("model/guarantees"), "model");
  assertEquals(groupOf("home"), null);
  assertEquals(aliasOf("commands/read"), "read");
  assertEquals(aliasOf("model/cst"), undefined);
  assertEquals(aliasOf("home"), undefined);
});

Deno.test("help catalog - lookups resolve and unknown names are undefined", () => {
  assertEquals(findByReach("read")?.id, "commands/read");
  assertEquals(findByReach("cst")?.id, "model/cst");
  assertEquals(findByAlias("read")?.id, "commands/read");
  assertEquals(findByAlias("cst"), undefined);
  assertEquals(findPage("concepts/selectors")?.id, "concepts/selectors");
  assertEquals(findPage("home")?.id, "home");
  assertEquals(findByReach("nope"), undefined);
  assertEquals(findByAlias("nope"), undefined);
  assertEquals(findPage("nope/nope"), undefined);
});

Deno.test("help catalog - grouped map covers all non-home pages", () => {
  const grouped = groupedPages();
  const groups = grouped.map((g) => g.group);
  assertEquals(groups, ["model", "concepts", "commands"]);
  const total = grouped.reduce((n, g) => n + g.pages.length, 0);
  assertEquals(total, HELP_PAGES.length - 1); // home is excluded
  // 2 model + 6 concepts + 9 commands
  assertEquals(grouped[0].pages.length, 2);
  assertEquals(grouped[1].pages.length, 6);
  assertEquals(grouped[2].pages.length, 9);
});

// --- Phase 1: content migration invariants (dev log 113) ---

Deno.test("help catalog - every furtherReading target resolves to a real page", () => {
  for (const p of HELP_PAGES) {
    for (const link of p.furtherReading) {
      assertEquals(
        findPage(link.page) !== undefined,
        true,
        `${p.id} links to unknown page '${link.page}'`,
      );
      assertEquals(link.hint.length > 0, true);
    }
  }
});

Deno.test("help catalog - every page has at least one non-empty section", () => {
  for (const p of HELP_PAGES) {
    assertEquals(p.sections.length > 0, true, `${p.id} has no sections`);
    for (const s of p.sections) {
      assertEquals(s.body.length > 0, true);
      if (s.heading.length === 0) {
        // Only the home intro renders bare (straight under the splash).
        assertEquals(p.id, "home", `${p.id} has an empty section heading`);
      }
    }
  }
});

Deno.test("help catalog - furtherReading never links to itself", () => {
  for (const p of HELP_PAGES) {
    for (const link of p.furtherReading) {
      assertEquals(link.page === p.id, false, `${p.id} links to itself`);
    }
  }
});

// --- Phase 3: renderers ---

Deno.test("help render - text is deterministic, marker-free, and non-empty", () => {
  for (const p of HELP_PAGES) {
    const text = renderPageText(p);
    assertEquals(text.length > 0, true, `${p.id} renders empty`);
    assertEquals(text.includes("\x1b["), false, `${p.id} has ANSI`);
    assertEquals(text.includes("`"), false, `${p.id} has stray backticks`);
  }
});

Deno.test("help render - rich=never and rich=auto (non-TTY) equal the text floor", () => {
  for (const p of HELP_PAGES) {
    assertEquals(renderPage(p, "never"), renderPageText(p), `${p.id} never != text`);
    assertEquals(renderPage(p, "auto"), renderPageText(p), `${p.id} auto (non-TTY) != text`);
  }
});

Deno.test("help render - rich=always is ANSI, stripped equals text, runs are balanced", () => {
  for (const p of HELP_PAGES) {
    if (p.id === "home") {
      // Home adds the rich-only splash, so the strip-equality below does not
      // hold; assert only the renderer-mode contract (ANSI in rich mode).
      assertEquals(renderPageRich(p).includes("\x1b["), true, "home rich has no ANSI");
      continue;
    }
    const rich = renderPageRich(p);
    assertEquals(rich.includes("\x1b["), true, `${p.id} has no ANSI`);
    assertEquals(stripAnsi(rich), renderPageText(p), `${p.id} rich != text after strip`);
    // deno-lint-ignore no-control-regex
    const escapes = rich.match(/\x1b\[[0-9;]*m/g) ?? [];
    for (let i = 0; i < escapes.length; i++) {
      if (escapes[i] !== "\x1b[0m") {
        assertEquals(escapes[i + 1], "\x1b[0m", `${p.id}: unbalanced ANSI after ${escapes[i]}`);
      }
    }
  }
});

File: tests/help_test.rs
//! Help catalog and renderer (Deno `tests/help_test.ts`).

use lq::{
    HELP_PAGES, HelpPage, PageGroup, RichMode, alias_of, find_by_alias, find_by_reach, find_page,
    group_of, grouped_pages, reach_of, render_page_rich, render_page_text, render_page_tty,
};

const EXPECTED_IDS: &[&str] = &[
    "home",
    "model/cst",
    "model/guarantees",
    "concepts/state-scope",
    "concepts/private-notes",
    "concepts/insets",
    "concepts/selectors",
    "concepts/mutations",
    "concepts/tracked-changes",
    "commands/init",
    "commands/schema",
    "commands/dump",
    "commands/preview",
    "commands/read",
    "commands/bib",
    "commands/set",
    "commands/delete",
    "commands/insert",
    "commands/undo",
];

fn strip_ansi(s: &str) -> String {
    let mut out = String::new();
    let mut rest = s;
    while let Some(idx) = rest.find('\u{1b}') {
        out.push_str(&rest[..idx]);
        rest = &rest[idx + 1..];
        if rest.starts_with('[')
            && let Some(end) = rest.find('m')
        {
            rest = &rest[end + 1..];
            continue;
        }
        out.push('\u{1b}');
    }
    out.push_str(rest);
    out
}

#[test]
fn page_map_matches_the_draft() {
    let mut ids: Vec<&str> = HELP_PAGES.iter().map(|p| p.id).collect();
    ids.sort_unstable();
    let mut expected = EXPECTED_IDS.to_vec();
    expected.sort_unstable();
    assert_eq!(ids, expected);
}

#[test]
fn page_ids_are_unique() {
    let ids: Vec<&str> = HELP_PAGES.iter().map(|p| p.id).collect();
    let mut uniq = ids.clone();
    uniq.sort_unstable();
    uniq.dedup();
    assert_eq!(uniq.len(), ids.len());
}

#[test]
fn reaches_are_unique() {
    let reaches: Vec<&str> = HELP_PAGES.iter().map(|p| reach_of(p.id)).collect();
    let mut uniq = reaches.clone();
    uniq.sort_unstable();
    uniq.dedup();
    assert_eq!(uniq.len(), reaches.len());
}

#[test]
fn command_aliases_are_unique_and_only_on_command_pages() {
    let aliases: Vec<&str> = HELP_PAGES.iter().filter_map(|p| alias_of(p.id)).collect();
    let mut uniq = aliases.clone();
    uniq.sort_unstable();
    uniq.dedup();
    assert_eq!(uniq.len(), aliases.len());
    assert_eq!(aliases.len(), 10);
}

#[test]
fn preview_help_lists_utf8_warning_classes() {
    let page = find_page("commands/preview").expect("preview page");
    let output = page
        .sections
        .iter()
        .find(|s| s.heading == "Output")
        .expect("Output")
        .body;
    assert!(output.contains("warnings"), "{output}");
    assert!(output.contains("UTF-8"), "{output}");
    assert!(output.contains("include"), "{output}");
}

#[test]
fn every_page_has_a_non_empty_title() {
    for p in HELP_PAGES {
        assert!(!p.title.is_empty(), "{}", p.id);
    }
}

#[test]
fn group_reach_alias_derived_from_id() {
    assert_eq!(reach_of("commands/read"), "read");
    assert_eq!(reach_of("model/cst"), "cst");
    assert_eq!(reach_of("home"), "home");
    assert_eq!(group_of("commands/read"), Some(PageGroup::Commands));
    assert_eq!(group_of("concepts/insets"), Some(PageGroup::Concepts));
    assert_eq!(group_of("model/guarantees"), Some(PageGroup::Model));
    assert_eq!(group_of("home"), None);
    assert_eq!(alias_of("commands/read"), Some("read"));
    assert_eq!(alias_of("model/cst"), None);
    assert_eq!(alias_of("home"), None);
}

#[test]
fn lookups_resolve() {
    assert_eq!(find_by_reach("read").map(|p| p.id), Some("commands/read"));
    assert_eq!(find_by_reach("cst").map(|p| p.id), Some("model/cst"));
    assert_eq!(find_by_alias("read").map(|p| p.id), Some("commands/read"));
    assert!(find_by_alias("cst").is_none());
    assert_eq!(
        find_page("concepts/selectors").map(|p| p.id),
        Some("concepts/selectors")
    );
    assert_eq!(find_page("home").map(|p| p.id), Some("home"));
    assert!(find_by_reach("nope").is_none());
    assert!(find_by_alias("nope").is_none());
    assert!(find_page("nope/nope").is_none());
}

#[test]
fn grouped_map_covers_non_home() {
    let grouped = grouped_pages();
    let groups: Vec<PageGroup> = grouped.iter().map(|g| g.group).collect();
    assert_eq!(
        groups,
        vec![PageGroup::Model, PageGroup::Concepts, PageGroup::Commands]
    );
    let total: usize = grouped.iter().map(|g| g.pages.len()).sum();
    assert_eq!(total, HELP_PAGES.len() - 1);
    assert_eq!(grouped[0].pages.len(), 2);
    assert_eq!(grouped[1].pages.len(), 6);
    assert_eq!(grouped[2].pages.len(), 10);
}

#[test]
fn further_reading_targets_resolve() {
    for p in HELP_PAGES {
        for link in p.further_reading {
            assert!(
                find_page(link.page).is_some(),
                "{} links to unknown page '{}'",
                p.id,
                link.page
            );
            assert!(!link.hint.is_empty());
        }
    }
}

#[test]
fn every_page_has_a_non_empty_section() {
    for p in HELP_PAGES {
        assert!(!p.sections.is_empty(), "{} has no sections", p.id);
        for s in p.sections {
            assert!(!s.body.is_empty());
            if s.heading.is_empty() {
                assert_eq!(p.id, "home", "{} has an empty section heading", p.id);
            }
        }
    }
}

#[test]
fn further_reading_never_links_to_self() {
    for p in HELP_PAGES {
        for link in p.further_reading {
            assert_ne!(link.page, p.id, "{} links to itself", p.id);
        }
    }
}

fn all_pages() -> impl Iterator<Item = &'static HelpPage> {
    HELP_PAGES.iter()
}

#[test]
fn text_is_deterministic_marker_free_non_empty() {
    for p in all_pages() {
        let text = render_page_text(p);
        assert!(!text.is_empty(), "{} renders empty", p.id);
        assert!(!text.contains("\x1b["), "{} has ANSI", p.id);
        assert!(!text.contains('`'), "{} has stray backticks", p.id);
    }
}

#[test]
fn rich_never_and_auto_non_tty_equal_text() {
    for p in all_pages() {
        assert_eq!(
            render_page_tty(p, RichMode::Never, false),
            render_page_text(p),
            "{} never != text",
            p.id
        );
        assert_eq!(
            render_page_tty(p, RichMode::Auto, false),
            render_page_text(p),
            "{} auto (non-TTY) != text",
            p.id
        );
    }
}

#[test]
fn rich_always_ansi_stripped_equals_text() {
    for p in all_pages() {
        if p.id == "home" {
            assert!(
                render_page_rich(p).contains("\x1b["),
                "home rich has no ANSI"
            );
            continue;
        }
        let rich = render_page_rich(p);
        assert!(rich.contains("\x1b["), "{} has no ANSI", p.id);
        assert_eq!(
            strip_ansi(&rich),
            render_page_text(p),
            "{} rich != text after strip",
            p.id
        );
        let escapes = collect_escapes(&rich);
        for i in 0..escapes.len() {
            if escapes[i] != "\x1b[0m" {
                assert_eq!(
                    escapes.get(i + 1).map(String::as_str),
                    Some("\x1b[0m"),
                    "{}: unbalanced ANSI after {}",
                    p.id,
                    escapes[i]
                );
            }
        }
    }
}

fn collect_escapes(s: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut rest = s;
    while let Some(idx) = rest.find('\u{1b}') {
        rest = &rest[idx..];
        if rest.starts_with("\x1b[")
            && let Some(end) = rest.find('m')
        {
            out.push(rest[..=end].to_string());
            rest = &rest[end + 1..];
            continue;
        }
        rest = &rest[1..];
    }
    out
}

#[test]
fn state_scope_help_names_undo_snapshot_warning() {
    let page = find_page("concepts/state-scope").expect("state-scope page");
    let body: String = page.sections.iter().map(|s| s.body).collect();
    assert!(
        body.contains(
            "When the snapshot cannot be saved, that mutation's JSON includes a warning."
        ),
        "{body}"
    );
}
